python/beta-estimation.qmd

```diff
@@ -0,0 +1,464 @@
+                 .dropna()
+                 .groupby('permno', group_keys=False)
+                 )
+
+# Calculate beta values using joblib parallel processing
+beta_monthly_joblib = (
+    pd.concat(joblib.Parallel(n_jobs=number_of_cores)
+              (joblib.delayed(roll_capm_estimation_for_joblib)(name, group)
+               for name, group in permno_groups)
+              )
+    .dropna()
+    )
+
+# To check if all the betas are the same:
+print((beta_monthly['beta_monthly'] - beta_monthly_joblib['beta']).sum())
+
+```
+
+## Estimating Beta using Daily Returns
+
+Before we provide some descriptive statistics of our beta estimates, we implement the estimation for the daily CRSP sample as well. Depending on the application, you might either use longer horizon beta estimates based on monthly data or shorter horizon estimates based on daily returns. 
+
+First, we load daily CRSP data. 
+Note that the sample is large compared to the monthly data, so make sure to have enough memory available.
+
+```{python}
+crsp_daily = (pd.read_sql_query(
+    sql="SELECT permno, month, date, ret_excess FROM crsp_daily",
+    con=tidy_finance,
+    parse_dates={"month": {"unit": "D","origin": "unix"},
+                 "date": {"unit": "D","origin": "unix"}})
+  .dropna()
+)
+```
+
+We also need the daily Fama-French market excess returns.
+
+```{python}
+factors_ff_daily = (pd.read_sql_query(
+    sql="SELECT date, mkt_excess FROM factors_ff_daily",
+    con=tidy_finance,
+    parse_dates={"date": {"unit": "D","origin": "unix"}})
+  .dropna()
+)
+```
+
+We make sure to keep only relevant data to save memory space. 
+However, note that your machine might not have enough memory to read the whole daily CRSP sample. In this case, we refer you to the exercises and try working with loops. 
+
+```{python}
+window_size = 90
+min_obs = 50
+
+valid_permnos = (crsp_daily
+  .groupby("permno")["permno"]
+  .count()
+  .reset_index(name="counts")
+  .query("counts > @window_size + 1")
+  .drop(columns = "counts")
+)
+
+crsp_daily = (crsp_daily
+  .merge(factors_ff_daily, 
+         how="inner", 
+         on="date")
+  .merge(valid_permnos, 
+         how="inner", 
+         on="permno")
+)
+```
+
+Even though we could now just apply the function using `groupby()` on the whole CRSP sample, we advise against doing it as it is computationally quite expensive. Remember that we have to perform rolling-window estimations across all stocks and time periods. However, this estimation problem is an ideal scenario to employ the power of parallelization. Parallelization means that we split the tasks which perform rolling-window estimations across different workers (or cores on your local machine).
+
+```{python}
+permno_groups = (crsp_daily
+                 .merge(valid_permnos, how="inner", on="permno")
+                 .dropna()
+                 .groupby('permno', group_keys=False)
+                 )
+
+# Calculate beta values using joblib parallel processing
+betas = (
```

Comment:
This object should be called beta_daily, not beta and ideally we only carry forward the last beta of each month (we never use daily betas and do not store them in the database).

Reply:
Implemented the changes myself